.cbmc-batch/jobs/aws_cryptosdk_enc_ctx_serialize/aws_cryptosdk_enc_ctx_serialize_harness.c

```diff
@@ -27,24 +27,28 @@ void array_list_item_generator(struct aws_array_list *elems) {
     for (size_t index = 0; index < elems->length; ++index) {
         struct aws_hash_element *val = (struct aws_hash_element *)((uint8_t *)elems->data + (elems->item_size * index));
         // Due to the checks in aws_cryptosdk_enc_ctx_size, no string can have a length > UINT16_MAX
-        val->key   = ensure_string_is_allocated_bounded_length(UINT16_MAX);
+        __CPROVER_assume(val != NULL);
```

Comment:
Don't we have an invariant for `aws_hash_element`?

Reply:
I couldn't find one in c-common, either in hash_table.h or make_common_data_structures or other cbmc source code.

Reply:
Different hash elements will have different constraints. What you need is one that ensures the particular requirements of this type.

Reply:
I'm concerned about this `val != NULL` assumption. val comes from `data` so if its `null`, something has gone wrong higher up in the code. I'd prefer to make this `assert(AWS_MEM_IS_WRITABLE(val, elems->item_size)`

Reply:
So I just tried out that assert and it fails. I think it's because we don't actually specify anything about the elements of the hash table, and we use this array_list_item_generator as a generator for those elements. I found this comment describing the strategy: 
```
 /* If the consumer of the list doesn't use the elements in the list, we can just leave it undef
 * This is sound, as it gives you a totally nondet value every time you use a list element, and is the default behaviour
 * of CBMC. But if it is used, we need a way for the harness to specify valid values for the element, for example if
 * they are copying values out of the table. They can do this by defining
 * -DAWS_CRYPTOSDK_HASH_ELEMS_ARRAY_INIT_GENERATOR=the_generator_fn, wher the_generator_fn has signature
 * the_generator_fn(struct aws_array_list *elems).
 *   [elems] is a pointer to the array_list whose values need to be set
 */
```

Reply:
If that assertion fails, that implies that whatever setup is being done for the array list is insufficient.

Reply:
I'm looking into this. The array list isn't set up as part of the harness but rather through aws_cryptosdk_hash_elems_array_init which we stub out (to handle CBMC function pointer removal). The stub for aws_cryptosdk_hash_elems_array_init uses the generator defined in the harness to nondet specify elements (with the assumptions from the harness on them). This is why the assert fails. 
I'm trying our replacing the stub with the actual function but elements are copied in via memset and other functions needed for the assertions to hold are also stubbed out, so it's a bit of a mess. Hopefully I can get something sorted out.